File: 17. Letter Combinations of a Phone Number/Solution.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        keys = {'2': ['a', 'b', 'c'], '3': ['d', 'e', 'f'], '4': ['g', 'h', 'i'], '5': ['j', 'k', 'l'], '6': ['m', 'n', 'o'], '7': ['p', 'q', 'r', 's'], '8': ['t', 'u', 'v'], '9': ['w', 'x', 'y', 'z']}
        if not digits: return []
        if len(digits) == 1: return keys[digits]
        res = []
        for k in keys[digits[0]]:
            for foll in self.letterCombinations(digits[1:]):
                res.append(k + foll)
        return res

File: 17. Letter Combinations of a Phone Number/Solution.java
class Solution {
    Map<Character, List<String>> keys = Map.of(
        '2', new ArrayList<>(List.of("a", "b", "c")),
        '3', new ArrayList<>(List.of("d", "e", "f")),
        '4', new ArrayList<>(List.of("g", "h", "i")),
        '5', new ArrayList<>(List.of("j", "k", "l")),
        '6', new ArrayList<>(List.of("m", "n", "o")),
        '7', new ArrayList<>(List.of("p", "q", "r", "s")),
        '8', new ArrayList<>(List.of("t", "u", "v")),
        '9', new ArrayList<>(List.of("w", "x", "y", "z"))
    );
    public List<String> letterCombinations(String digits) {
        if (digits.length() == 0) return new ArrayList<>();
        if (digits.length() == 1) return keys.get(digits.charAt(0));
        List<String> res = new ArrayList<>();
        for (String k : keys.get(digits.charAt(0))) {
            for (String foll : letterCombinations(digits.substring(1))) {
                res.add(k + foll);
            }
        }
        return res;
    }
}
